### DISCUSSION TITLE
pilotnet_snn slayer training - gradient is not available for output layer CUBA LIF neuron's voltage_state

### DISCUSSION THREAD

**Discussion by gwgknudayanga** [lava-nc/lava-dl#362] [Q&A]
Hi,

In pilotnet snn, it seems the final logit values are achieved by reading the output layer neurons' voltage_states (in netx loihi runs).
However how the loss.backward() is calulated based on neuron.voltage_state values as it is neuron.voltage_state is not a parameter ? I can see it is in buffers, however when i set it's requires_grad = True, it is not set.

**Comment by bamsumit** [comment]
Why can you not call `current, voltage = neuron.dynamics(input)` and use the voltage tensor to calculate whatever loss you want? Are the gradient not flowing?

**Comment by gwgknudayanga** [comment]
@bamsumit 
* I tried that as well as the following.
     output = torch.sum(self.blocks[4].neuron.voltage_state,(2,3)) + torch.matmul(torch.zeros(10).to(self.weights.device), self.weights)
* But in any of those cases network was not learning. It learns only when when use the output of the network output's layer.

* So i used spikingjelly to learn and then converted that to slayer and inference on slayer. In that case, inferencing in slayer shows 95% of test accuracy. However when i deployed this in loihi2 it only 60% of accuracy. I used quantize_8bit function during training for weights and meanonlybatchnorm. I cannot understand why this much of accuracy drop when implemeted in loihi2.
